#### In this notebook we will run collaborative filtering methods, namely item item and user user, and determine the best among both and use that for final recommendation

In [1]:
import pandas as pd
import re,string
import numpy as np

- We had removed the null values of important columns and saved it as df.csv, hence we are importing the same

In [2]:
df = pd.read_csv("df.csv")
df.head()

Unnamed: 0            brand  \
0           0  Universal Music   
1           1         Lundberg   
2           2         Lundberg   
3           3              K-Y   
4           4              K-Y   

                                          categories  \
0  Movies, Music & Books,Music,R&b,Movies & TV,Mo...   
1  Food,Packaged Foods,Snacks,Crackers,Snacks, Co...   
2  Food,Packaged Foods,Snacks,Crackers,Snacks, Co...   
3  Personal Care,Medicine Cabinet,Lubricant/Sperm...   
4  Personal Care,Medicine Cabinet,Lubricant/Sperm...   

                         manufacturer  \
0  Universal Music Group / Cash Money   
1                            Lundberg   
2                            Lundberg   
3                                 K-Y   
4                                 K-Y   

                                         name reviews_didPurchase  \
0   Pink Friday: Roman Reloaded Re-Up (w/dvd)                 NaN   
1  Lundberg Organic Cinnamon Toast Rice Cakes                True   
2  Lundberg Organic Cinnamon Toast Rice Cakes                True   
3            K-Y Love Sensuality Pleasure Gel               False   
4            K-Y Love Sensuality Pleasure Gel               False   

  reviews_doRecommend  reviews_rating  \
0                 NaN               5   
1                 NaN               5   
2                 NaN               5   
3               False               1   
4               False               1   

                                        reviews_text reviews_title  \
0  i love this album. it's very good. more to the...  Just Awesome   
1  Good flavor. This review was collected as part...          Good   
2                                       Good flavor.          Good   
3  I read through the reviews on here before look...  Disappointed   
4  My husband bought this gel for us. The gel cau...    Irritation   

  reviews_userCity reviews_userProvince reviews_username user_sentiment  \
0      Los Angeles                  NaN           joshua       Positive   
1              NaN                  NaN        dorothy w       Positive   
2              NaN                  NaN        dorothy w       Positive   
3              NaN                  NaN          rebecca       Negative   
4              NaN                  NaN        walker557       Negative   

                                  reviews_title_text  
0  Just Awesomei love this album. it's very good....  
1  GoodGood flavor. This review was collected as ...  
2                                   GoodGood flavor.  
3  DisappointedI read through the reviews on here...  
4  IrritationMy husband bought this gel for us. T...

- We see that there are no null username in this df

In [3]:
df[df.reviews_username.isna()]

Empty DataFrame
Columns: [Unnamed: 0, brand, categories, manufacturer, name, reviews_didPurchase, reviews_doRecommend, reviews_rating, reviews_text, reviews_title, reviews_userCity, reviews_userProvince, reviews_username, user_sentiment, reviews_title_text]
Index: []

In [4]:
df_for_recommendation = df


- Since there are duplicates for 'reviews_username' and 'name', hence we shall remove it. I found there are duplictated because pivot table could not be created previously.

In [5]:
df_for_recommendation = df_for_recommendation.drop_duplicates(subset=['reviews_username','name'],keep='first')

- Splitting data between train and test

In [6]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(df_for_recommendation, test_size=0.30, random_state=42)

- Creating pivot to understand relation between reviews_username and product name

In [7]:
df_pivot = train.pivot(
    index='reviews_username',
    columns='name',
    values='reviews_rating').fillna(0)
df_pivot.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         0.0               
02dakota                                                        0.0               
02deuce                                                         0.0               
0325home                                                        0.0               
06stidriver                                                     0.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                           0.0   
02dakota                                          0.0   
02deuce                                           0.0   
0325home                                          0.0   
06stidriver                                       0.0   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         0.0                   
02dakota                                                        0.0                   
02deuce                                                         0.0                   
0325home                                                        0.0                   
06stidriver                                                     0.0                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              0.0   
02dakota                                             0.0   
02deuce                                              0.0   
0325home                                             0.0   
06stidriver                                          0.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     0.0   
02dakota                                                    0.0   
02deuce                                                     0.0   
0325home                                                    0.0   
06stidriver                                                 0.0   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00sab00                                                         0.0   
02dakota                                                        0.0   
02deuce                                                         0.0   
0325home                                                        0.0   
06stidriver                                                     0.0   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00sab00                                                         0.0             
02dakota                                                        0.0             
02deuce                                                         0.0             
0325home                                                        0.0             
06stidriver                                                     0.0             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00sab00                                                         0.0    
02dakota                                                        0.0    
02deuce                                                         0.0    
0325home                                                        0.0    
06stidriver        

### Creating dummy train & dummy test dataset


In [8]:

dummy_train = train.copy()
dummy_train['reviews_rating'] = dummy_train['reviews_rating'].apply(lambda x: 0 if x>0 else 1)

In [9]:
dummy_train = dummy_train.pivot(
    index='reviews_username',
    columns='name',
    values='reviews_rating'
).fillna(1)

In [10]:
dummy_train.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         1.0               
02dakota                                                        1.0               
02deuce                                                         1.0               
0325home                                                        1.0               
06stidriver                                                     1.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                           1.0   
02dakota                                          1.0   
02deuce                                           1.0   
0325home                                          1.0   
06stidriver                                       1.0   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         1.0                   
02dakota                                                        1.0                   
02deuce                                                         1.0                   
0325home                                                        1.0                   
06stidriver                                                     1.0                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              1.0   
02dakota                                             1.0   
02deuce                                              1.0   
0325home                                             1.0   
06stidriver                                          1.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     1.0   
02dakota                                                    1.0   
02deuce                                                     1.0   
0325home                                                    1.0   
06stidriver                                                 1.0   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00sab00                                                         1.0   
02dakota                                                        1.0   
02deuce                                                         1.0   
0325home                                                        1.0   
06stidriver                                                     1.0   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00sab00                                                         1.0             
02dakota                                                        1.0             
02deuce                                                         1.0             
0325home                                                        1.0             
06stidriver                                                     1.0             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00sab00                                                         1.0    
02dakota                                                        1.0    
02deuce                                                         1.0    
0325home                                                        1.0    
06stidriver        

In [11]:
df_pivot = train.pivot(
    index='reviews_username',
    columns='name',
    values='reviews_rating'
)
df_pivot.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         NaN               
02dakota                                                        NaN               
02deuce                                                         NaN               
0325home                                                        NaN               
06stidriver                                                     NaN               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                           NaN   
02dakota                                          NaN   
02deuce                                           NaN   
0325home                                          NaN   
06stidriver                                       NaN   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         NaN                   
02dakota                                                        NaN                   
02deuce                                                         NaN                   
0325home                                                        NaN                   
06stidriver                                                     NaN                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              NaN   
02dakota                                             NaN   
02deuce                                              NaN   
0325home                                             NaN   
06stidriver                                          NaN   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     NaN   
02dakota                                                    NaN   
02deuce                                                     NaN   
0325home                                                    NaN   
06stidriver                                                 NaN   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00sab00                                                         NaN   
02dakota                                                        NaN   
02deuce                                                         NaN   
0325home                                                        NaN   
06stidriver                                                     NaN   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00sab00                                                         NaN             
02dakota                                                        NaN             
02deuce                                                         NaN             
0325home                                                        NaN             
06stidriver                                                     NaN             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00sab00                                                         NaN    
02dakota                                                        NaN    
02deuce                                                         NaN    
0325home                                                        NaN    
06stidriver        

- Using adjusted cosine similarity

In [12]:

mean = np.nanmean(df_pivot, axis=1)
df_subtracted = (df_pivot.T-mean).T

In [13]:
df_subtracted.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         NaN               
02dakota                                                        NaN               
02deuce                                                         NaN               
0325home                                                        NaN               
06stidriver                                                     NaN               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                           NaN   
02dakota                                          NaN   
02deuce                                           NaN   
0325home                                          NaN   
06stidriver                                       NaN   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         NaN                   
02dakota                                                        NaN                   
02deuce                                                         NaN                   
0325home                                                        NaN                   
06stidriver                                                     NaN                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              NaN   
02dakota                                             NaN   
02deuce                                              NaN   
0325home                                             NaN   
06stidriver                                          NaN   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     NaN   
02dakota                                                    NaN   
02deuce                                                     NaN   
0325home                                                    NaN   
06stidriver                                                 NaN   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00sab00                                                         NaN   
02dakota                                                        NaN   
02deuce                                                         NaN   
0325home                                                        NaN   
06stidriver                                                     NaN   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00sab00                                                         NaN             
02dakota                                                        NaN             
02deuce                                                         NaN             
0325home                                                        NaN             
06stidriver                                                     NaN             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00sab00                                                         NaN    
02dakota                                                        NaN    
02deuce                                                         NaN    
0325home                                                        NaN    
06stidriver        

In [14]:
from sklearn.metrics.pairwise import pairwise_distances

In [15]:
# Creating the User Similarity Matrix using pairwise_distance function.
user_correlation = 1 - pairwise_distances(df_subtracted.fillna(0), metric='cosine')
user_correlation[np.isnan(user_correlation)] = 0
print(user_correlation)

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


#### Predicting user-user based similarity.

- Ignoring correlation <0 since it means users are not interested in it

In [16]:
user_correlation[user_correlation<0]=0
user_correlation

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [17]:
user_predicted_ratings = np.dot(user_correlation, df_pivot.fillna(0))
user_predicted_ratings

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [18]:
user_predicted_ratings.shape

(17732, 237)

- Since we are interested only in the products not rated by the user, we will ignore the products rated by the user by making it zero.

In [19]:
user_final_rating = np.multiply(user_predicted_ratings,dummy_train)
user_final_rating.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         0.0               
02dakota                                                        0.0               
02deuce                                                         0.0               
0325home                                                        0.0               
06stidriver                                                     0.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                           0.0   
02dakota                                          0.0   
02deuce                                           0.0   
0325home                                          0.0   
06stidriver                                       0.0   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         0.0                   
02dakota                                                        0.0                   
02deuce                                                         0.0                   
0325home                                                        0.0                   
06stidriver                                                     0.0                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              0.0   
02dakota                                             0.0   
02deuce                                              0.0   
0325home                                             0.0   
06stidriver                                          0.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     0.0   
02dakota                                                    0.0   
02deuce                                                     0.0   
0325home                                                    0.0   
06stidriver                                                 0.0   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00sab00                                                         0.0   
02dakota                                                        0.0   
02deuce                                                         0.0   
0325home                                                        0.0   
06stidriver                                                     0.0   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00sab00                                                         0.0             
02dakota                                                        0.0             
02deuce                                                         0.0             
0325home                                                        0.0             
06stidriver                                                     0.0             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00sab00                                                         0.0    
02dakota                                                        0.0    
02deuce                                                         0.0    
0325home                                                        0.0    
06stidriver        

#### Finding top 5 products

In [20]:
user_input = input("Enter your user name")
print(user_input)

Enter your user namejoshua
joshua


In [21]:
d = user_final_rating.loc[user_input].sort_values(ascending=False)[0:5]

In [22]:
d

name
Planes: Fire Rescue (2 Discs) (includes Digital Copy) (blu-Ray/dvd)    3.933900
Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total              2.665009
Bounce Dryer Sheets, Fresh Linen, 160 sheets                           2.665009
Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd        2.139524
The Resident Evil Collection 5 Discs (blu-Ray)                         1.688657
Name: joshua, dtype: float64

#### Evaltuation for User-User based similarity

In [23]:

common = test[test.reviews_username.isin(train.reviews_username)]
common.shape

(1040, 15)

In [24]:
common.head()

Unnamed: 0                 brand  \
20800       20984              Tostitos   
2594         2619  Universal Home Video   
11442       11540                Clorox   
9025         9119                Clorox   
4363         4412                Hoover   

                                              categories manufacturer  \
20800  Food,Packaged Foods,Snacks,Chips & Pretzels,Fo...     Tostitos   
2594   Movies, Music & Books,Movies,Comedy,Movies & T...    Universal   
11442  Household Essentials,Cleaning Supplies,Kitchen...       Clorox   
9025   Household Essentials,Cleaning Supplies,Kitchen...       Clorox   
4363   Home,Home Appliances,Vacuums & Floor Care,Upri...       Hoover   

                                                    name reviews_didPurchase  \
20800                  Tostitos Bite Size Tortilla Chips               False   
2594   My Big Fat Greek Wedding 2 (blu-Ray + Dvd + Di...                 NaN   
11442  Clorox Disinfecting Wipes Value Pack Scented 1...               False   
9025   Clorox Disinfecting Wipes Value Pack Scented 1...                 NaN   
4363   Hoover174 Platinum Collection153 Lightweight B...                 NaN   

      reviews_doRecommend  reviews_rating  \
20800                True               5   
2594                False               3   
11442                True               5   
9025                 True               5   
4363                  NaN               5   

                                            reviews_text  \
20800                                            Awesome   
2594   The first was definitely better than this sequ...   
11442  Love the smell of this one and Clorox is strea...   
9025   We donated these to our daughters classroom an...   
4363   Suction is superb. Brings the nap up in the ca...   

                        reviews_title reviews_userCity reviews_userProvince  \
20800                             Yum              NaN                  NaN   
2594   Definitely Second to the First              NaN                  NaN   
11442             streak free shine!!              NaN                  NaN   
9025     Works Great At Killing Germs         Murrieta                  NaN   
4363    Best vacuum I have ever owned              NaN                  NaN   

      reviews_username user_sentiment  \
20800           momof3       Positive   
2594             jbasa       Positive   
11442            mshoe       Positive   
9025         gtguru497       Positive   
4363            donnap       Positive   

                                      reviews_title_text  
20800                                         YumAwesome  
2594   Definitely Second to the FirstThe first was de...  
11442  streak free shine!!Love the smell of this one ...  
9025   Works Great At Killing GermsWe donated these t...  
4363   Best vacuum I have ever ownedSuction is superb...

In [25]:
common_user_based_matrix = common.pivot_table(index='reviews_username', columns='name', values='reviews_rating')

In [26]:
user_correlation_df = pd.DataFrame(user_correlation)

In [27]:
df_subtracted.head(1)

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         NaN               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                           NaN   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         NaN                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              NaN   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     NaN   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00sab00                                                         NaN   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00sab00                                                         NaN             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00sab00                                                         NaN    

name              Alex Cross (dvdvideo)  \
reviews_username                          
00sab00                             NaN   

name              All,bran Complete Wheat Flakes, 18 Oz.  ...  \
reviews_username                                          ...   
00sab00                                              NaN  ...   

name              Voortman Sugar Free Fudge Chocolate Chip Cookies  \
reviews_username                                                     
00sab00                                                        NaN   

name              Wagan Smartac 80watt Inverter With Usb  \
reviews_username                                           
00sab00                                              NaN   

name              Walkers Stem Ginger Shortbread  \
reviews_username                                   
00sab00                                      NaN   

name              Way Basics 3-Shelf Eco Narrow Bookcase Storage Shelf, Espresso - Formaldehyde Free - Lifetime Guarantee  \
reviews_username                                                                                                            
00sab00                                                         NaN                                                         

name              Wedding Wishes Wedding Guest Book  Weleda Everon Lip Balm  \
reviews_username                                                              
00sab00                                         NaN                     NaN   

name              Wilton Black Dots Standard Baking Cups  \
reviews_username                                           
00sab00                                              NaN   

name              Windex Original Glass Cleaner Refill 67.6oz (2 Liter)  \
reviews_username                                                          
00sab00                                                         NaN       

name              Yes To Carrots Nourishing Body Wash  \
reviews_username                                        
00sab00                                           NaN   

name              Yes To Grapefruit Rejuvenating Body Wash  
reviews_username                                            
00sab00                                                NaN  

[1 row

In [28]:
user_correlation_df['reviews_username'] = df_subtracted.index
user_correlation_df.set_index('reviews_username',inplace=True)
user_correlation_df.head()

0      1      2      3      4      5      6      7      \
reviews_username                                                           
00sab00             0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
02dakota            0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
02deuce             0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
0325home            0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
06stidriver         0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

                  8      9      ...  17722  17723  17724  17725  17726  17727  \
reviews_username                ...                                             
00sab00             0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
02dakota            0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
02deuce             0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
0325home            0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   
06stidriver         0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   

                  17728  17729  17730  17731  
reviews_username                              
00sab00             0.0    0.0    0.0    0.0  
02dakota            0.0    0.0    0.0    0.0  
02deuce             0.0    0.0    0.0    0.0  
0325home            0.0    0.0    0.0    0.0  
06stidriver         0.0    0.0    0.0    0.0  

[5 rows x 17732 columns]

In [29]:
list_name = common.reviews_username.tolist()

user_correlation_df.columns = df_subtracted.index.tolist()


user_correlation_df_1 =  user_correlation_df[user_correlation_df.index.isin(list_name)]

In [30]:
user_correlation_df_1.shape




(859, 17732)

In [31]:
user_correlation_df_2 = user_correlation_df_1.T[user_correlation_df_1.T.index.isin(list_name)]

user_correlation_df_3 = user_correlation_df_2.T

user_correlation_df_3.head()



00sab00  1234  1witch  37f5p  abby  acjuarez08  \
reviews_username                                                   
00sab00               0.0   0.0     0.0    0.0   0.0         0.0   
1234                  0.0   0.0     0.0    0.0   0.0         0.0   
1witch                0.0   0.0     0.0    0.0   0.0         0.0   
37f5p                 0.0   0.0     0.0    0.0   0.0         0.0   
abby                  0.0   0.0     0.0    0.0   0.0         0.0   

                  actionaction  adam  adelynsmom  aimee  ...  wolverine  \
reviews_username                                         ...              
00sab00                    0.0   0.0         0.0    0.0  ...        0.0   
1234                       0.0   0.0         0.0    0.0  ...        0.0   
1witch                     0.0   0.0         0.0    0.0  ...        0.0   
37f5p                      0.0   0.0         0.0    0.0  ...        0.0   
abby                       0.0   0.0         0.0    0.0  ...        0.0   

                  woottos  wvshopaholic  xavier  xmom  yeyo  yohnie1  yummy  \
reviews_username                                                              
00sab00               0.0           0.0     0.0   0.0   0.0      0.0    0.0   
1234                  0.0           0.0     0.0   0.0   0.0      0.0    0.0   
1witch                0.0           0.0     0.0   0.0   0.0      0.0    0.0   
37f5p                 0.0           0.0     0.0   0.0   0.0      0.0    0.0   
abby                  0.0           0.0     0.0   0.0   0.0      0.0    0.0   

                  zach  zmom  
reviews_username              
00sab00            0.0   0.0  
1234               0.0   0.0  
1witch             0.0   0.0  
37f5p              0.0   0.0  
abby               0.0   0.0  

[5 rows x 859 columns]

In [32]:
user_correlation_df_3.shape

user_correlation_df_3[user_correlation_df_3<0]=0



In [33]:
common_user_based_matrix

name              100:Complete First Season (blu-Ray)  Alex Cross (dvdvideo)  \
reviews_username                                                               
00sab00                                           NaN                    NaN   
1234                                              NaN                    NaN   
1witch                                            NaN                    NaN   
37f5p                                             5.0                    NaN   
abby                                              NaN                    NaN   
...                                               ...                    ...   
yeyo                                              NaN                    NaN   
yohnie1                                           NaN                    NaN   
yummy                                             NaN                    NaN   
zach                                              NaN                    NaN   
zmom                                              NaN                    NaN   

name              All,bran Complete Wheat Flakes, 18 Oz.  \
reviews_username                                           
00sab00                                              NaN   
1234                                                 NaN   
1witch                                               NaN   
37f5p                                                NaN   
abby                                                 NaN   
...                                                  ...   
yeyo                                                 NaN   
yohnie1                                              NaN   
yummy                                                NaN   
zach                                                 NaN   
zmom                                                 NaN   

name              Ambi Complexion Cleansing Bar  \
reviews_username                                  
00sab00                                     NaN   
1234                                        NaN   
1witch                                      NaN   
37f5p                                       NaN   
abby                                        NaN   
...                                         ...   
yeyo                                        NaN   
yohnie1                                     NaN   
yummy                                       NaN   
zach                                        NaN   
zmom                                        NaN   

name              Arrid Extra Dry Anti-Perspirant Deodorant Spray Regular  \
reviews_username                                                            
00sab00                                                         NaN         
1234                                                            NaN         
1witch                                                          NaN         
37f5p                                                           NaN         
abby                                                            NaN         
...                                                             ...         
yeyo                                                            NaN         
yohnie1                                                         NaN         
yummy                                                           NaN         
zach                                                            NaN         
zmom                                                            NaN         

name              Aussie Aussome Volume Shampoo, 13.5 Oz  \
reviews_username                                           
00sab00                                              NaN   
1234                                                 NaN   
1witch                                               NaN   
37f5p                                                NaN   
abby                                                 NaN   
...                                                  ...   
yeyo                            

In [34]:
common_user_predicted_ratings = np.dot(user_correlation_df_3, common_user_based_matrix.fillna(0))
common_user_predicted_ratings




array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [35]:
dummy_test = common.copy()
dummy_test['reviews_rating'] = dummy_test['reviews_rating'].apply(lambda x: 1 if x>0 else 0)

dummy_test = dummy_test.pivot_table( index='reviews_username',
    columns='name',
    values='reviews_rating').fillna(0)

print(dummy_test.shape)

common_user_predicted_ratings = np.multiply(common_user_predicted_ratings,dummy_test)

(859, 112)


In [36]:
common_user_predicted_ratings.head(2)

name              100:Complete First Season (blu-Ray)  Alex Cross (dvdvideo)  \
reviews_username                                                               
00sab00                                           0.0                    0.0   
1234                                              0.0                    0.0   

name              All,bran Complete Wheat Flakes, 18 Oz.  \
reviews_username                                           
00sab00                                              0.0   
1234                                                 0.0   

name              Ambi Complexion Cleansing Bar  \
reviews_username                                  
00sab00                                     0.0   
1234                                        0.0   

name              Arrid Extra Dry Anti-Perspirant Deodorant Spray Regular  \
reviews_username                                                            
00sab00                                                         0.0         
1234                                                            0.0         

name              Aussie Aussome Volume Shampoo, 13.5 Oz  \
reviews_username                                           
00sab00                                              0.0   
1234                                                 0.0   

name              Aveeno Baby Continuous Protection Lotion Sunscreen with Broad Spectrum SPF 55, 4oz  \
reviews_username                                                                                       
00sab00                                                         0.0                                    
1234                                                            0.0                                    

name              Avery174 Ready Index Contemporary Table Of Contents Divider, 1-8, Multi, Letter  \
reviews_username                                                                                    
00sab00                                                         0.0                                 
1234                                                            0.0                                 

name              Banana Boat Sunless Summer Color Self Tanning Lotion, Light To Medium  \
reviews_username                                                                          
00sab00                                                         0.0                       
1234                                                            0.0                       

name              Bisquick Original Pancake And Baking Mix - 40oz  ...  \
reviews_username                                                   ...   
00sab00                                                       0.0  ...   
1234                                                          0.0  ...   

name              Various Artists - Choo Choo Soul (cd)  \
reviews_username                                          
00sab00                                             0.0   
1234                                                0.0   

name              Vaseline Intensive Care Healthy Hands Stronger Nails  \
reviews_username                                                         
00sab00                                                         0.0      
1234                                                            0.0      

name              Vaseline Intensive Care Lip Therapy Cocoa Butter  \
reviews_username                                                     
00sab00                                                        0.0   
1234                                                           0.0   

name              Vicks Vaporub, Regular, 3.53oz  \
reviews_username                                   
00sab00                                      0.0   
1234                                         0.0   

name              Voortman Sugar Free Fudge Chocolate Chip Cookies  \
reviews_username                                                     
00sab00                            

#### Calculating the RMSE for only the movies rated by user.

In [37]:
from sklearn.preprocessing import MinMaxScaler
from numpy import *

X  = common_user_predicted_ratings.copy() 
X = X[X>0]

scaler = MinMaxScaler(feature_range=(1, 5))
print(scaler.fit(X))
y = (scaler.transform(X))

print(y)

MinMaxScaler(feature_range=(1, 5))
[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]


C:\Users\pbabbar\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:373: RuntimeWarning: All-NaN slice encountered
  data_min = np.nanmin(X, axis=0)
C:\Users\pbabbar\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:374: RuntimeWarning: All-NaN slice encountered
  data_max = np.nanmax(X, axis=0)


In [38]:
common_ = common.pivot_table( index='reviews_username',
    columns='name',
    values='reviews_rating')

In [39]:
total_non_nan = np.count_nonzero(~np.isnan(y))

In [40]:
rmse = (sum(sum((common_ - y )**2))/total_non_nan)**0.5
print(round(rmse,2))

2.52


RMSE for user user is 2.52

### Item based similarity

In [42]:
df_pivot = train.pivot(
   index='name',
    columns='reviews_username',
    values='reviews_rating'
)

df_pivot.head()

[[1. 0. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 0. 1.]]


In [ ]:
mean = np.nanmean(df_pivot, axis=1)
df_subtracted = (df_pivot.T-mean).T

df_subtracted.head()

In [ ]:
from sklearn.metrics.pairwise import pairwise_distances

# Item Similarity Matrix
item_correlation = 1 - pairwise_distances(df_subtracted.fillna(0), metric='cosine')
item_correlation[np.isnan(item_correlation)] = 0
print(item_correlation)

In [43]:
item_correlation[item_correlation<0]=0
item_correlation

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [ ]:
#### Predition Item Item

In [44]:

item_predicted_ratings = np.dot((df_pivot.fillna(0).T),item_correlation)
item_predicted_ratings
item_predicted_ratings.shape
dummy_train.shape



(17732, 237)

#### Filetering products not rated by user

In [45]:
item_final_rating = np.multiply(item_predicted_ratings,dummy_train)
item_final_rating.head()

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         0.0               
02dakota                                                        0.0               
02deuce                                                         0.0               
0325home                                                        0.0               
06stidriver                                                     0.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                      0.029419   
02dakota                                     0.007697   
02deuce                                      0.006158   
0325home                                     0.000000   
06stidriver                                  0.000000   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         0.0                   
02dakota                                                        0.0                   
02deuce                                                         0.0                   
0325home                                                        0.0                   
06stidriver                                                     0.0                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              0.0   
02dakota                                             0.0   
02deuce                                              0.0   
0325home                                             0.0   
06stidriver                                          0.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     0.0   
02dakota                                                    0.0   
02deuce                                                     0.0   
0325home                                                    0.0   
06stidriver                                                 0.0   

name              4C Grated Parmesan Cheese 100% Natural 8oz Shaker  \
reviews_username                                                      
00sab00                                                         0.0   
02dakota                                                        0.0   
02deuce                                                         0.0   
0325home                                                        0.0   
06stidriver                                                     0.0   

name              Africa's Best No-Lye Dual Conditioning Relaxer System Super  \
reviews_username                                                                
00sab00                                                         0.0             
02dakota                                                        0.0             
02deuce                                                         0.0             
0325home                                                        0.0             
06stidriver                                                     0.0             

name              Alberto VO5 Salon Series Smooth Plus Sleek Shampoo  \
reviews_username                                                       
00sab00                                                         0.0    
02dakota                                                        0.0    
02deuce                                                         0.0    
0325home                                                        0.0    
06stidriver        

#### Finding top 5 recommendations

In [46]:
user_input = input("Enter your user name")
print(user_input)

Enter your user namejoshua
joshua


In [47]:
d = item_final_rating.loc[user_input].sort_values(ascending=False)[0:5]
d

name
Chester's Cheese Flavored Puffcorn Snacks          0.166022
Nexxus Exxtra Gel Style Creation Sculptor          0.060284
Tostitos Bite Size Tortilla Chips                  0.056611
Cococare 100% Natural Castor Oil                   0.045543
Diet Canada Dry Ginger Ale - 12pk/12 Fl Oz Cans    0.043133
Name: joshua, dtype: float64

#### Evaluation - Item Item based recommendation

In [48]:
test.columns

common =  test[test.name.isin(train.name)]
common.shape

common.head(4)

common_item_based_matrix = common.pivot_table(index='name', columns='reviews_username', values='reviews_rating')

common_item_based_matrix.shape

item_correlation_df = pd.DataFrame(item_correlation)

item_correlation_df.head(1)

item_correlation_df['name'] = df_subtracted.index
item_correlation_df.set_index('name',inplace=True)
item_correlation_df.head()

0    1    2    3    4    \
name                                                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  1.0  0.0  0.0  0.0  0.0   
100:Complete First Season (blu-Ray)                 0.0  1.0  0.0  0.0  0.0   
2017-2018 Brownline174 Duraflex 14-Month Planne...  0.0  0.0  0.0  0.0  0.0   
2x Ultra Era with Oxi Booster, 50fl oz              0.0  0.0  0.0  1.0  0.0   
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  0.0  0.0  0.0  0.0   

                                                    5    6    7    8    9    \
name                                                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  0.0  0.0  0.0  0.0  0.0   
100:Complete First Season (blu-Ray)                 0.0  0.0  0.0  0.0  0.0   
2017-2018 Brownline174 Duraflex 14-Month Planne...  0.0  0.0  0.0  0.0  0.0   
2x Ultra Era with Oxi Booster, 50fl oz              0.0  0.0  0.0  0.0  0.0   
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  0.0  0.0  0.0  0.0   

                                                    ...  227  228  229  230  \
name                                                ...                       
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  ...  0.0  0.0  0.0  0.0   
100:Complete First Season (blu-Ray)                 ...  0.0  0.0  0.0  0.0   
2017-2018 Brownline174 Duraflex 14-Month Planne...  ...  0.0  0.0  0.0  0.0   
2x Ultra Era with Oxi Booster, 50fl oz              ...  0.0  0.0  0.0  0.0   
42 Dual Drop Leaf Table with 2 Madrid Chairs"       ...  0.0  0.0  0.0  0.0   

                                                    231  232  233  234  235  \
name                                                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  0.0  0.0  0.0  0.0  0.0   
100:Complete First Season (blu-Ray)                 0.0  0.0  0.0  0.0  0.0   
2017-2018 Brownline174 Duraflex 14-Month Planne...  0.0  0.0  0.0  0.0  0.0   
2x Ultra Era with Oxi Booster, 50fl oz              0.0  0.0  0.0  0.0  0.0   
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  0.0  0.0  0.0  0.0   

                                                    236  
name                                                     
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  0.0  
100:Complete First Season (blu-Ray)                 0.0  
2017-2018 Brownline174 Duraflex 14-Month Planne...  0.0  
2x Ultra Era with Oxi Booster, 50fl oz              0.0  
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  

[5 rows x 237 columns]

In [49]:

list_name = common.name.tolist()
item_correlation_df.columns = df_subtracted.index.tolist()
item_correlation_df_1 =  item_correlation_df[item_correlation_df.index.isin(list_name)]
item_correlation_df_2 = item_correlation_df_1.T[item_correlation_df_1.T.index.isin(list_name)]

In [50]:
item_correlation_df_3 = item_correlation_df_2.T



In [51]:
item_correlation_df_3.head()

item_correlation_df_3[item_correlation_df_3<0]=0



In [52]:
common_item_predicted_ratings = np.dot(item_correlation_df_3, common_item_based_matrix.fillna(0))
common_item_predicted_ratings


array([[0.        , 0.00749298, 0.00461828, ..., 0.        , 0.01367573,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.01661788, 0.00073382, 0.01110969, ..., 0.0030172 , 0.        ,
        0.        ],
       [0.        , 0.        , 0.00381293, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [53]:
dummy_test = common.copy()

dummy_test['reviews_rating'] = dummy_test['reviews_rating'].apply(lambda x: 1 if x>0 else 0)

dummy_test = dummy_test.pivot_table(index='name', columns='reviews_username', values='reviews_rating').fillna(0)

common_item_predicted_ratings = np.multiply(common_item_predicted_ratings,dummy_test)

In [54]:
common_ = common.pivot_table(index='name', columns='reviews_username', values='reviews_rating')

In [55]:
from sklearn.preprocessing import MinMaxScaler
from numpy import *

X  = common_item_predicted_ratings.copy() 
X = X[X>0]

scaler = MinMaxScaler(feature_range=(1, 5))
print(scaler.fit(X))
y = (scaler.transform(X))

print(y)

MinMaxScaler(feature_range=(1, 5))
[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]


C:\Users\pbabbar\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:373: RuntimeWarning: All-NaN slice encountered
  data_min = np.nanmin(X, axis=0)
C:\Users\pbabbar\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:374: RuntimeWarning: All-NaN slice encountered
  data_max = np.nanmax(X, axis=0)


In [56]:
total_non_nan = np.count_nonzero(~np.isnan(y))

In [57]:
rmse = (sum(sum((common_ - y )**2))/total_non_nan)**0.5
print(rmse)

3.5740059611894166


- Since rmse is less for user-user, we will use user user rating.

In [59]:
item_final_rating

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         0.0               
02dakota                                                        0.0               
02deuce                                                         0.0               
0325home                                                        0.0               
06stidriver                                                     0.0               
...                                                             ...               
zwithanx                                                        0.0               
zxcsdfd                                                         0.0               
zxjki                                                           0.0               
zzdiane                                                         0.0               
zzz1127                                                         0.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                      0.029419   
02dakota                                     0.007697   
02deuce                                      0.006158   
0325home                                     0.000000   
06stidriver                                  0.000000   
...                                               ...   
zwithanx                                     0.000000   
zxcsdfd                                      0.000000   
zxjki                                        0.000000   
zzdiane                                      0.049032   
zzz1127                                      0.000000   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         0.0                   
02dakota                                                        0.0                   
02deuce                                                         0.0                   
0325home                                                        0.0                   
06stidriver                                                     0.0                   
...                                                             ...                   
zwithanx                                                        0.0                   
zxcsdfd                                                         0.0                   
zxjki                                                           0.0                   
zzdiane                                                         0.0                   
zzz1127                                                         0.0                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              0.0   
02dakota                                             0.0   
02deuce                                              0.0   
0325home                                             0.0   
06stidriver                                          0.0   
...                                                  ...   
zwithanx                                             0.0   
zxcsdfd                                              0.0   
zxjki                                                0.0   
zzdiane                                              0.0   
zzz1127                                              0.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     0.0   
02dakota                                                    0.0 

In [60]:
user_final_rating

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         0.0               
02dakota                                                        0.0               
02deuce                                                         0.0               
0325home                                                        0.0               
06stidriver                                                     0.0               
...                                                             ...               
zwithanx                                                        0.0               
zxcsdfd                                                         0.0               
zxjki                                                           0.0               
zzdiane                                                         0.0               
zzz1127                                                         0.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                           0.0   
02dakota                                          0.0   
02deuce                                           0.0   
0325home                                          0.0   
06stidriver                                       0.0   
...                                               ...   
zwithanx                                          0.0   
zxcsdfd                                           0.0   
zxjki                                             0.0   
zzdiane                                           0.0   
zzz1127                                           0.0   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         0.0                   
02dakota                                                        0.0                   
02deuce                                                         0.0                   
0325home                                                        0.0                   
06stidriver                                                     0.0                   
...                                                             ...                   
zwithanx                                                        0.0                   
zxcsdfd                                                         0.0                   
zxjki                                                           0.0                   
zzdiane                                                         0.0                   
zzz1127                                                         0.0                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              0.0   
02dakota                                             0.0   
02deuce                                              0.0   
0325home                                             0.0   
06stidriver                                          0.0   
...                                                  ...   
zwithanx                                             0.0   
zxcsdfd                                              0.0   
zxjki                                                0.0   
zzdiane                                              0.0   
zzz1127                                              0.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     0.0   
02dakota                                                    0.0 

In [61]:
item_final_rating.loc['joshua'].sort_values(ascending=False)[0:5]

name
Chester's Cheese Flavored Puffcorn Snacks          0.166022
Nexxus Exxtra Gel Style Creation Sculptor          0.060284
Tostitos Bite Size Tortilla Chips                  0.056611
Cococare 100% Natural Castor Oil                   0.045543
Diet Canada Dry Ginger Ale - 12pk/12 Fl Oz Cans    0.043133
Name: joshua, dtype: float64

In [62]:
user_final_rating.loc['joshua'].sort_values(ascending=False)[0:5]

name
Planes: Fire Rescue (2 Discs) (includes Digital Copy) (blu-Ray/dvd)    3.933900
Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total              2.665009
Bounce Dryer Sheets, Fresh Linen, 160 sheets                           2.665009
Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd        2.139524
The Resident Evil Collection 5 Discs (blu-Ray)                         1.688657
Name: joshua, dtype: float64

In [63]:
user_final_rating

name              0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  \
reviews_username                                                                  
00sab00                                                         0.0               
02dakota                                                        0.0               
02deuce                                                         0.0               
0325home                                                        0.0               
06stidriver                                                     0.0               
...                                                             ...               
zwithanx                                                        0.0               
zxcsdfd                                                         0.0               
zxjki                                                           0.0               
zzdiane                                                         0.0               
zzz1127                                                         0.0               

name              100:Complete First Season (blu-Ray)  \
reviews_username                                        
00sab00                                           0.0   
02dakota                                          0.0   
02deuce                                           0.0   
0325home                                          0.0   
06stidriver                                       0.0   
...                                               ...   
zwithanx                                          0.0   
zxcsdfd                                           0.0   
zxjki                                             0.0   
zzdiane                                           0.0   
zzz1127                                           0.0   

name              2017-2018 Brownline174 Duraflex 14-Month Planner 8 1/2 X 11 Black  \
reviews_username                                                                      
00sab00                                                         0.0                   
02dakota                                                        0.0                   
02deuce                                                         0.0                   
0325home                                                        0.0                   
06stidriver                                                     0.0                   
...                                                             ...                   
zwithanx                                                        0.0                   
zxcsdfd                                                         0.0                   
zxjki                                                           0.0                   
zzdiane                                                         0.0                   
zzz1127                                                         0.0                   

name              2x Ultra Era with Oxi Booster, 50fl oz  \
reviews_username                                           
00sab00                                              0.0   
02dakota                                             0.0   
02deuce                                              0.0   
0325home                                             0.0   
06stidriver                                          0.0   
...                                                  ...   
zwithanx                                             0.0   
zxcsdfd                                              0.0   
zxjki                                                0.0   
zzdiane                                              0.0   
zzz1127                                              0.0   

name              42 Dual Drop Leaf Table with 2 Madrid Chairs"  \
reviews_username                                                  
00sab00                                                     0.0   
02dakota                                                    0.0 

In [64]:
user_final_rating.to_csv("user_final_rating.csv")